app/controllers/review_mapping_controller.rb

```diff
@@ -51,129 +47,196 @@ def select_metareviewer
-            assignment_team = begin
-                                assignment_teams.to_a.sample
-                              rescue StandardError
-                                nil
-                              end
-            if assignment_team.nil?
-              flash[:error] = 'No artifacts are available to review at this time. Please try later.'
-            else
-              assignment.assign_reviewer_dynamically_no_topic(reviewer, assignment_team)
-            end
-          end
-        else
-          flash[:error] = 'You cannot do more reviews when you have ' + Assignment.max_outstanding_reviews + 'reviews to do'
-        end
-      else
-        flash[:error] = 'You cannot do more than ' + assignment.num_reviews_allowed.to_s + ' reviews based on assignment policy'
-      end
-      # rescue Exception => e
-      #   flash[:error] = (e.nil?) ? $! : e
-      # end
+      assign_reviewer_based_on_topics(assignment, reviewer)
     end
+
+    flash[:error] = error_message if error_message
+
     redirect_to controller: 'student_review', action: 'list', id: participant.id
   end
 
-  # This method checks if the user is allowed to do any more reviews.
-  # First we find the number of reviews done by that reviewer for that assignment and we compare it with assignment policy
-  # if number of reviews are less than allowed than a user is allowed to request.
-  def review_allowed?(assignment, reviewer)
-    @review_mappings = ReviewResponseMap.where(reviewer_id: reviewer.id, reviewed_object_id: assignment.id)
-    assignment.num_reviews_allowed > @review_mappings.size
-  end
 
-  # This method checks if the user that is requesting a review has any outstanding reviews, if a user has more than 2
-  # outstanding reviews, he is not allowed to ask for more reviews.
-  # First we find the reviews done by that student, if he hasn't done any review till now, true is returned
-  # else we compute total reviews completed by adding each response
-  # we then check of the reviews in progress are less than assignment's policy
-  def check_outstanding_reviews?(assignment, reviewer)
-    @review_mappings = ReviewResponseMap.where(reviewer_id: reviewer.id, reviewed_object_id: assignment.id)
-    @num_reviews_total = @review_mappings.size
-    if @num_reviews_total.zero?
-      true
-    else
-      @num_reviews_completed = 0
-      @review_mappings.each do |map|
-        @num_reviews_completed += 1 if !map.response.empty? && map.response.last.is_submitted
-      end
-      @num_reviews_in_progress = @num_reviews_total - @num_reviews_completed
-      @num_reviews_in_progress < Assignment.max_outstanding_reviews
-    end
-  end
+
 
   # assigns the quiz dynamically to the participant
   def assign_quiz_dynamically
     begin
-      assignment = Assignment.find(params[:assignment_id])
-      reviewer = AssignmentParticipant.where(user_id: params[:reviewer_id], parent_id: assignment.id).first
+      assignment = find_assignment(params[:assignment_id])
```

Comment:
Why is a method call being substituted for a direct invocation of a db function?  This seems to make the code more obscure instead of more readable.  I think this should be reverted.